packages/firestore/test/unit/specs/existence_filter_spec.test.ts

```diff
@@ -286,4 +246,347 @@ describeSpec('Existence Filters:', [], () => {
       );
     }
   );
+
+  /**
+   * Test existence filter with bloom filter.
+   */
+  specTest(
+    'Full re-query is skipped when bloom filter can identify documents deleted',
+    [],
+    () => {
+      const query1 = query('collection');
+      const docA = doc('collection/a', 1000, { v: 1 });
+      const docB = doc('collection/b', 1000, { v: 2 });
+      const bloomFilterProto = generateBloomFilterProto({
+        contains: [docA],
+        notContains: [docB],
+        numOfBits: 2,
+        hashCount: 1
+      });
+      return (
+        spec()
+          .userListens(query1)
+          .watchAcksFull(query1, 1000, docA, docB)
+          .expectEvents(query1, { added: [docA, docB] })
+          // DocB is deleted in the next sync.
+          .watchFilters([query1], [docA.key], bloomFilterProto)
+          .watchSnapshots(2000)
+          // BloomFilter identify docB is deleted, skip full query and put docB
+          // into limbo directly.
+          .expectEvents(query1, { fromCache: true })
+          .expectLimboDocs(docB.key) // DocB is now in limbo.
+          .ackLimbo(2000, deletedDoc('collection/b', 2000))
+          .expectLimboDocs() // DocB is no longer in limbo.
+          .expectEvents(query1, {
+            removed: [docB]
+          })
+      );
+    }
+  );
+
+  specTest(
+    'Full re-query is triggered when bloom filter can not identify documents deleted',
+    [],
+    () => {
+      const query1 = query('collection');
+      const docA = doc('collection/a', 1000, { v: 1 });
+      const docB = doc('collection/b', 1000, { v: 2 });
+      const docC = doc('collection/c', 1000, { v: 2 });
+      const bloomFilterProto = generateBloomFilterProto({
+        contains: [docA, docB],
+        notContains: [docC],
+        numOfBits: 3,
+        hashCount: 1
+      });
+      return (
+        spec()
+          .userListens(query1)
+          .watchAcksFull(query1, 1000, docA, docB, docC)
+          .expectEvents(query1, { added: [docA, docB, docC] })
+          // DocB and DocC are deleted in the next sync.
+          .watchFilters([query1], [docA.key], bloomFilterProto)
+          .watchSnapshots(2000)
+          // BloomFilter correctly identifies docC is deleted, but yields false
+          // positive results for docB. Re-run query is triggered.
+          .expectEvents(query1, { fromCache: true })
+          .expectActiveTargets({ query: query1, resumeToken: '' })
+          .watchRemoves(query1) // Acks removal of query.
```

Comment:
Is it possible to move the call to `.watchRemoves(query1)` to earlier in the test? Its presence here is a bit confusing because it looks like the target is removed, then it gets a result in the next line. For example, could it be moved to _above_ the `.expectEvents(query1, { fromCache: true })` line?

Reply:
WatchSnapshots is supposed to be followed by expectEvents as they will be write into one event while building.  I can move it one line up, above "expectActiveTargets"

Reply:
I removed the re-query and limbo resolution steps, as they are irrelevant to the Bloom Filter application. Only checking we have triggered re-query, and have an active target of query1 with empty resume token.